src/grab.ts

```diff
@@ -1,157 +1,121 @@
+        }
         return acc;
-      }, []);
+      }, {});
 
-      projections.push(`...select(${condProjections.join(", ")})`);
-    }
+    // Split base and conditional fields
+    const conditionalFields = Object.values(
+      conditionalSelections || {}
+    ) as Selection[];
 
-    // Schema gets a bit trickier, since we sort of have to mock GROQ behavior.
-    const schema = (() => {
-      const toSchemaInput = (sel: Selection) =>
-        Object.entries(sel).reduce<z.ZodRawShape>((acc, [key, value]) => {
-          if ("schema" in value) {
-            acc[key] = value.schema;
-          } else if (Array.isArray(value)) {
-            acc[key] = value[1];
-          } else {
-            acc[key] = value;
-          }
-          return acc;
-        }, {});
+    const baseSchema = z.object(toSchemaInput(selection));
+    const conditionalFieldSchemas = conditionalFields.map((field) =>
+      baseSchema.merge(z.object(toSchemaInput(field)))
+    );
 
-      // Unknown schema means we just use the selection passed
-      if (
-        (prev.schema instanceof z.ZodArray
-          ? prev.schema.element
-          : prev.schema) instanceof z.ZodUnknown
-      ) {
-        // Split base and conditional fields
-        const conditionalFields = Object.values(
-          conditionalSelections || {}
-        ) as Selection[];
+    const unionEls = [...conditionalFieldSchemas, baseSchema];
+    const s =
+      unionEls.length > 1
+        ? z.union([unionEls[0], unionEls[1], ...unionEls.slice(2)])
+        : unionEls[0];
 
-        const baseSchema = z.object(toSchemaInput(selection));
-        const conditionalFieldSchemas = conditionalFields.map((field) =>
-          baseSchema.merge(z.object(toSchemaInput(field)))
-        );
-        const s =
-          conditionalFieldSchemas.length === 0
-            ? baseSchema
-            : // eslint-disable-next-line @typescript-eslint/ban-ts-comment
-              // @ts-ignore Need to figure out how to make tsc happy
-              z.union([...conditionalFieldSchemas, baseSchema]);
+    return schema instanceof z.ZodArray ? z.array(s) : s;
+  })();
 
-        return prev.schema instanceof z.ZodArray ? z.array(s) : s;
-      }
-      // If we're already dealing with an object schema inside our array, we need to do a Pick
-      else if (
-        (prev.schema instanceof z.ZodArray
-          ? prev.schema.element
-          : prev.schema) instanceof z.ZodObject
-      ) {
-        const toPick = Object.keys(selection).reduce<{
-          [key: string]: true;
-        }>((acc, key) => {
-          acc[key] = true;
-          return acc;
-        }, {});
+  const res = (newSchema instanceof z.ZodArray
+    ? new ArrayQuery({
+        query: query + `{${projections.join(", ")}}`,
+        schema: newSchema,
+      })
+    : new EntityQuery({
+        query: query + `{${projections.join(", ")}}`,
+        schema: newSchema,
+      })) as unknown as ArrayQuery<AllSelection>;
```

Comment:
```suggestion
      })) as unknown as NewType;
```
Then you can omit the cast on the return statement.